src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -53,20 +53,48 @@ private static bool IsEnabled()
                    s_diagnosticListener.IsEnabled();
         }
 
-        private static Activity? CreateActivity(HttpRequestMessage requestMessage)
+        private static Activity? StartActivity(HttpRequestMessage request)
         {
             Activity? activity = null;
             if (s_activitySource.HasListeners())
             {
-                activity = s_activitySource.CreateActivity(DiagnosticsHandlerLoggingStrings.ActivityName, ActivityKind.Client);
+                Uri? requestUri = request.RequestUri;
+                if (requestUri?.IsAbsoluteUri == false)
+                {
+                    requestUri = null;
+                }
+
+                // The following tags might be important for making sampling decisions and should be provided at span creation time.
```

Comment:
> Do you know of any example of a sampler that looks at these tags to make a sampling decision?

I haven't seen a real one.

Reply:
it's quite popular in Java - https://learn.microsoft.com/en-us/azure/azure-monitor/app/java-standalone-sampling-overrides.

One of the motivations to provide them at start time is that the same attributes are populated for metrics and you could optimize by reusing them between signals.
The spec doesn't REQUIRE to populate them, perf is one of the good reasons to not follow the recommendation. 

AFAIK selective sampling is usually used to drop uninteresting operations like HEAD requests or healthchecks. 
On .NET if I remember correctly, the common way to do it is via the processor. This is not a common/documented mechanism in OTel though.

-----
TL;DR

Semconv does not block you from not providing attributes at creation time and users can rely on .NET-specific processor approach to drop uninteresting spans. 

I'm also quite interested in OTel .NET maintainers opinion on this.

Reply:
a bit more context:
- AWS X-Ray sampling relies on some of the attributes: https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-configuring-xray-sampling-rules.html
- it's also popular to do tail-based sampling and sample interesting stuff later - e.g. https://github.com/honeycombio/refinery

Reply:
@vishweshbankwar given that you were also pushing for this, can you provide some insights why did you consider it important?

Reply:
+1 to Noah's concern. For client spans, I haven't seen asks to get the tags at sampler time. Have seen the ask for server -asp.net and asp.net core quite often - eg: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1938 and https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1871. Probably because Server typically *makes* the decision while HttpClient follows that of the parent.

If it is cheap to provide these tags at creation time - its great. But from current Activity API, this can't be done without allocations, so I think its best to provide Tags *after* decision is known for client

(For DataBase client span - we do provide one tag at creation time - but that is a statically known tag : https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlActivitySourceHelper.cs#L25-L28)

Reply:
> can you provide some insights why did you consider it important

Mainly for spec compliance.

This requirement started with MUST which later was changed to SHOULD in the spec. https://github.com/open-telemetry/semantic-conventions/issues/422

Agree on the concerns related to the allocation costs. If https://github.com/dotnet/runtime/issues/103245 lands, then we could consider adding this later.

Reply:
Before we drop this, I would like to also note that Activity tags are stored in a linked list, which inherently leads to allocation overhead, see #83345. This means we have ~8 alloactions for the tag list already, so having +1 doesn't really matter much, unless we believe we will address that problem in reasonable time. The only scenario where it would matter is when the sampling decision would be `PropagationData`, so we don't have to populate the tags.

Reply:
Thanks for all the feedback everyone!

> I wonder if we should commit to these bits of performance improvement work for .NET 10 instead of dismissing the standards recommendation?

Once we make these pre-sampling there some amount of perf overhead that would be difficult to ever get rid of even if we did the optimizations you were suggesting. Quick guess maybe we could reduce the overhead by 50% with more work. On the other hand if we add the tags after sampling right now we'll have 0 overhead from this and if anyone raises the lack of these tags for sampling as an issue we could always reconsider and make them available in the future.

Overall it sounds like there is no clear use-case to do head sampling with these tags right now. The xray example @lmolkova mentioned sounded the most likely but the current OTel instrumentation library for HttpClient doesn't populate these tags until after the activity is started and presumably if that was causing a problem for the xray sampler then the OTel.NET maintainers would be aware of it. It also sounds like the spec has already softened their stance around this issue.

> I would like to also note that Activity tags are stored in a linked list

My concern on perf overhead is about the performance of the case where Activity winds up not being created. For example if someone wants to sample 1% of requests then 1% of the time they pay that Activity linked list tags cost, but 99% of the time they are paying the cost to setup these tags and give them to the sampler, then discard them.

I'd recommend we create these tags after the Activity is created, then wait for feedback. For anyone who wants to do sampling on this data in the meantime any form of tail sampling remains an option for them and they are no worse off than they were in .NET 8.